### PR TITLE
[#12194] Add simple workflow to vet pull requests

### LINKED ISSUE #12194: Add simple workflow to vet pull requests
Ever since teammates-bot went down, maintainers have been manually checking if PR titles and descriptions adhere to our contribution guidelines. Given that Github Actions is commonly used for pull request management nowadays, let's take the opportunity to switch to a Github Actions workflow for checking PRs.

### PR DESCRIPTION
**Pull request by zhaojj2209**
Fixes #12194
Part of #11383

Adds simple Github Actions workflow using [actions/github-script](https://github.com/actions/github-script) to check PRs. The bot currently checks for the following:
- Title: starts with issue number in square brackets, followed by a space
- Description: contains either "Fixes" or "Part of", followed by the issue number. Only "Fixes" is checked since it is the keyword used in our PR template.

More checks can be added down the line as and when needed. The bot also only checks PRs that are newly opened/reopened, so as to avoid spam.

### CODE REVIEW

**Review by samuelfangjw — COMMENTED** [review]
Seems like this doesn't check the title/pr after it is edited. Is this something we would want?

**Comment by zhaojj2209** [comment]
> Seems like this doesn't check the title/pr after it is edited. Is this something we would want?

Github Actions has an `edited` type for PRs which I believe is triggered on editing PR title/body, but I intentionally left it out. Back when the old bot was active, there was a case where a contributor opened a PR then proceeded to spam title/description edits, resulting in maintainers watching the repo being spammed with comment notifications. There is therefore a trade-off between either higher workload for the maintainers (if the action does not rerun on edit) or higher risk of spam (if the action does rerun on edit).

@damithc @wkurniawan07 Would appreciate your input on this!

**Comment by samuelfangjw** [comment]
> Github Actions has an `edited` type for PRs which I believe is triggered on editing PR title/body, but I intentionally left it out. Back when the old bot was active, there was a case where a contributor opened a PR then proceeded to spam title/description edits, resulting in maintainers watching the repo being spammed with comment notifications. There is therefore a trade-off between either higher workload for the maintainers (if the action does not rerun on edit) or higher risk of spam (if the action does rerun on edit).

Have similar thoughts, my opinion is it's ok to leave it as such. Maintainers can still manually verify/edit if there are issues.

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          if (isTitleValid && isDescriptionValid) {
+            return
+          }
+          let body = `Hi @${pr.data.user.login}, thank you for your interest in contributing to TEAMMATES!
+            However, your PR does not appear to follow our [contribution guidelines](https://teammates.github.io/teammates/process.html#step-4-submit-a-pr):\n\n`
+          if (!isTitleValid) {
+            body += "- Title must start with the issue number the PR is fixing in square brackets, e.g. `[#<issue-number>]`\n"
+          }
+          if (!isDescriptionValid) {
+            body += "- Description must reference the issue number the PR is fixing, e.g. `Fixes #<issue-number>` (or `Part of #<issue-number>` if the PR does not addres the issue fully)\n"
```
Comment (samuelfangjw) [inline .github/workflows/pr.yml:33]:
```suggestion
            body += "- Description must reference the issue number the PR is fixing, e.g. `Fixes #<issue-number>` (or `Part of #<issue-number>` if the PR does not address the issue fully)\n"
```

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

**Review by samuelfangjw — APPROVED** [review]

**Comment by wkurniawan07** [comment]
@zhaojj2209 I have no strong opinion on rerun on edit. As you have identified, either one has its own set of pros/cons.

> Only "Fixes" is checked since it is the keyword used in our PR template.

The old bot checks for [all 9 acceptable keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue), so it's better to retain the behaviour here.

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    steps:
+    - uses: actions/github-script@v6
+      with:
+        script: |
+          const pr = await github.rest.pulls.get({
+            owner: context.repo.owner,
+            repo: context.repo.repo,
+            pull_number: context.issue.number
+          })
+          const isTitleValid = /^\[#\d\] /.test(pr.data.title)
```
Comment (hhdqirui) [inline .github/workflows/pr.yml:22]: There might be some issues with the regex.
```suggestion
          const isTitleValid = /^\[#\d+\] /.test(pr.data.title)
```

**Review by hhdqirui — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yml (hunk 2 above)
Comment (zhaojj2209) [reply .github/workflows/pr.yml:22]: Thanks for the catch!

**Review by hhdqirui — APPROVED** [review]

**Pull request merged**